### DISCUSSION TITLE
Keyboard shortcut for "Mute myself" with the space bar

### DISCUSSION THREAD

**Discussion by fjd** [jamulussoftware/jamulus#1693] [Ideas]
For singers it would be nice to have a quickly accessible key to mute their own mic. Known in German as _Räuspertaste_.

The spacebar is often used for this or it can be configured by yourself.

In addition, support for the headset button would also be possible here.

**Comment by vimpostor** [comment]
When you click on the "Mute" button, it will be focused. After that you can already press the "Space button" to mute/unmute. This will work, as long as the "Mute" button has focus.

Right now the "Mute" button does not have focus by default. The "Connect" button has focus by default, which makes sense to me. However, as soon as one connects to the server, it could make sense to focus the "Mute" button by default. Perhaps something like that could be implemented.

**Reply by fjd** [reply]
Unfortunately, this does not work with macOS.
There is no shortcut for mute myself and the idea with the focus and the space bar does not work either.

**Comment by gilgongo** [comment]
> In addition, support for the headset button would also be possible here.

Sorry, could you explain what you mean by "headset button"?

**Reply by hoffie** [reply]
Some headsets (or cables) have physical mute buttons and I assume that those are not implemented within the headsets themselves (at least when using non-persistent buttons on non-USB headsets), but rather by sending a signal to the operating system (which Jamulus may not currently support). Guest guessing though. :o

**Reply by fjd** [reply]
The last time I developed with QT was 20 years ago.
Maybe one of these keycodes will help?

e.g.
Qt::Key_ToggleCallHangup (0x01100007)
Qt::Key_MediaTogglePlayPause (0x1000086)

https://doc.qt.io/qt-5/qt.html

**Reply by hoffie** [reply]
That might be worth a try, yes. It would be best if someone with such a headset could experiment with the required code.

**Reply by henkdegroot** [reply]
Interesting sounds like a nice feature....guess the Qt::Key_MicMute is might make more sense to use. I do have a plantronics USB headset, so should be able to test. Don't have a real mac though....

EDIT: Sorry but can't test this as it already works! When I click the Mute button on the headset, the mic is already de-activated and does not sent any sound to Jamulus.

**Reply by ann0see** [reply]
Probably that's a hardware feature?

**Reply by henkdegroot** [reply]
> that's a hardware feature?

I think, it might be very hardware dependent.  Of course when I use the button, this does not replicate the "Mute Myself" state and technically it is even different as "Mute Myself" does not mute your input but just does not sent your input to the server.

**Comment by hoffie** [comment]
> For singers it would be nice to have a quickly accessible key to mute their own mic. Known in German as Räuspertaste.

The "Mute myself" checkbox does have an accelerator key, which is `M` at least in German (you can tell by the underlined M in `StuMmschalten`). So, Alt+M should already do what you want -- at least it does for me. Can you try that?
Didn't know the German Räuspertaste, btw. :D

**Reply by fjd** [reply]
Unfortunately not with macOS 😬 
The shortcut was forgotten here. The space bar is faster and more accessible than a key combination.

**Reply by hoffie** [reply]
> Unfortunately not with macOS

Hm, I'm wondering why that is though. The accelerator keys are part of the translations and translations are platform-independent. Do other accelerator keys in Jamulus work? Do accelerator keys in other applications work?

**Reply by npostavs** [reply]
Doesn't mac have different modifier keys? Maybe it should be Option+M instead of Alt+M, or something like that.

**Reply by hoffie** [reply]
Just found this in the [Qt docs](https://doc.qt.io/qt-5/qshortcut.html#details):

> On Mac, shortcuts are disabled by default. Call qt_set_sequence_auto_mnemonic() to enable them. However, because mnemonic shortcuts do not fit in with Aqua's guidelines, Qt will not show the shortcut character underlined.

We should probably consider activating them on Mac, too?
cc @softins @ann0see

**Reply by ann0see** [reply]
No problem with that. Once my macOS install works again I can test it

**Reply by hoffie** [reply]
Have opened #1726 to track this part (independent of the judgement whether that would be sufficient to solve this issue)

**Reply by henkdegroot** [reply]
Coding wise, this is not really complex to implement (i.e. hit space to toggle the Mute Myself when the dialog has the focus).
Easy to do and can even be custom to for a specific OS. However something to consider is the fact that others OS do respond on the "spacebar" when you have selected the "Mute Myself" (windows at least). There are more items in the dialog though that respond to "spacebar" when it is the selected item (as long as the dialog has the focus).

I would love to see this implemented (even though I am mainly a windows user) and could do this as well.

**Reply by hoffie** [reply]
> There are more items in the dialog though that respond to "spacebar" when it is the selected item (as long as the dialog has the focus).

I agree. We'll also have to ensure that this is window-specific. Thinking of typing text in the chat... :)

#1726 should be something which can be done fairly easily and could go into a possible 3.8.1 release.

**Reply by henkdegroot** [reply]
> that this is window-specific.

Well that is why I mentioned the dialog has the focus. As soon as another dialog (like chat) is opened, the spacebar no longer is a shortcut as the shortcut key is only defined for the clientdialog. So that part is covered.
The only thing you would "loose" is that when you have ticked another item in the GUI (i.e. a Solo or Mute option for a fader) you can no longer "untick" with space.
I have already a shortcut setup for this in my personal build, but using "NumPad 0" for this. Which allows me to have an external NumPad keyboard very close to me so I can easily "Mute Myself" without having to reach to my PC keyboard or mouse.
